chain/bitcoind_client.go

```diff
@@ -981,36 +1031,54 @@ func (c *BitcoindClient) FilterBlocks(
 
 	blockFilterer := NewBlockFilterer(c.chainConn.cfg.ChainParams, req)
 
-	// Iterate over the requested blocks, fetching each from the rpc client.
-	// Each block will scanned using the reverse addresses indexes generated
-	// above, breaking out early if any addresses are found.
-	for i, block := range req.Blocks {
-		// TODO(conner): add prefetching, since we already know we'll be
-		// fetching *every* block
-		rawBlock, err := c.GetBlock(&block.Hash)
+	blockHashes := make([]*chainhash.Hash, 0, len(req.Blocks))
+	for i := range req.Blocks {
+		blockHashes = append(blockHashes, &req.Blocks[i].Hash)
+	}
+
+	// Now we fetch blocks in interval of 15 to avoid out of memory errors
+	// in case of fetching too many blocks with GetBlocksBatch().
+	const batchSize = 15
+	total := len(blockHashes)
+	for i := 0; i < total; i += batchSize {
+		start := i
+		end := i + batchSize
+		if end > total {
+			end = total
+		}
+
+		rawBlocks, err := c.chainConn.GetBlocksBatch(
+			blockHashes[start:end],
+		)
 		if err != nil {
 			return nil, err
 		}
 
-		if !blockFilterer.FilterBlock(rawBlock) {
-			continue
-		}
+		// Iterate over the requested blocks, fetching each from the rpc
+		// client. Each block will scanned using the reverse addresses
+		// indexes generated above, breaking out early if any addresses
+		// are found.
+		for j := range rawBlocks {
+			if !blockFilterer.FilterBlock(rawBlocks[j]) {
+				continue
+			}
 
-		// If any external or internal addresses were detected in this
-		// block, we return them to the caller so that the rescan
-		// windows can widened with subsequent addresses. The
-		// `BatchIndex` is returned so that the caller can compute the
-		// *next* block from which to begin again.
-		resp := &FilterBlocksResponse{
-			BatchIndex:         uint32(i),
-			BlockMeta:          block,
-			FoundExternalAddrs: blockFilterer.FoundExternal,
-			FoundInternalAddrs: blockFilterer.FoundInternal,
-			FoundOutPoints:     blockFilterer.FoundOutPoints,
-			RelevantTxns:       blockFilterer.RelevantTxns,
-		}
+			// If any external or internal addresses were detected
+			// in this block, we return them to the caller so that
+			// the rescan windows can widened with subsequent addresses.
+			// The `BatchIndex` is returned so that the caller can
+			// compute the *next* block from which to begin again.
+			resp := &FilterBlocksResponse{
+				BatchIndex:         uint32(j + start),
```

Comment:
nit: maybe `start + j` is more logical here since `start` is the offset, and `j` is the index in the batch.